api/server/docker.go

```diff
@@ -218,30 +222,60 @@ func (d *driver) handshake(w http.ResponseWriter, r *http.Request) {
 }
 
 func (d *driver) attachToken(ctx context.Context, request *volumeRequest) (context.Context, string) {
-	token, tokenInName := d.GetTokenFromString(request.Name)
-	if !tokenInName {
-		token = request.Opts[api.Token]
+
+	// get token from name
+	tokenFromName, tokenInName := d.GetTokenFromString(request.Name)
+	if tokenInName {
+		return addTokenMetadata(ctx, tokenFromName), tokenFromName
 	}
-	if len(token) == 0 {
-		return ctx, ""
+
+	// get token from opts
+	tokenFromOpts := request.Opts[api.Token]
+	if tokenFromOpts != "" {
+		return addTokenMetadata(ctx, tokenFromOpts), tokenFromOpts
 	}
 
+	// get token secret
+	secret, context, ok := d.GetTokenSecretFromString(request.Name)
+	if ok && d.secretsStore != nil {
+		token, err := d.secretsStore.GetToken(secret, context)
+		if err != nil {
+			return ctx, ""
+		}
+
+		return addTokenMetadata(ctx, token), token
+	}
+
+	return ctx, ""
+}
+
+// addTokenMetadata adds the token to a given context's metadata
+func addTokenMetadata(ctx context.Context, token string) context.Context {
 	md := metadata.New(map[string]string{
 		"authorization": "bearer " + token,
 	})
-	return metadata.NewOutgoingContext(ctx, md), token
+	return metadata.NewOutgoingContext(ctx, md)
 }
 
 func (d *driver) attachTokenMount(ctx context.Context, request *mountRequest) (context.Context, string) {
-	token, _ := d.GetTokenFromString(request.Name)
-	if len(token) == 0 {
-		return ctx, ""
+	// get token from name
+	tokenFromName, tokenInName := d.GetTokenFromString(request.Name)
+	if tokenInName {
+		return addTokenMetadata(ctx, tokenFromName), tokenFromName
 	}
 
-	md := metadata.New(map[string]string{
-		"authorization": "bearer " + token,
-	})
-	return metadata.NewOutgoingContext(ctx, md), token
+	// get token secret
```

Comment:
nit: moving the GetTokenFromString and GetTokenSecretFromString to addTokenMetadata function, will reduce code duplication

Reply:
Ended up adding a `parseTokenInput` that's shared by `attachTokenMount` and `attachToken` to reduce code duplication